core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -31,256 +31,207 @@
-  private static final Logger log = Logger.getLogger(AbstractServerStream.class.getName());
-
-  /** Whether listener.closed() has been called. */
-  private boolean listenerClosed;
-  private ServerStreamListener listener;
-
-  private boolean headersSent = false;
+public abstract class AbstractServerStream extends AbstractStream2
+    implements ServerStream, MessageFramer.Sink {
   /**
-   * Whether the stream was closed gracefully by the application (vs. a transport-level failure).
+   * Only called from application thread. A separate interface just to avoid name
+   * collisions/confusion.
    */
-  private boolean gracefulClose;
-  /** Saved trailers from close() that need to be sent once the framer has sent all messages. */
-  private Metadata stashedTrailers;
+  protected interface Sink {
+    /**
+     * Sends response headers to the remote end points.
+     *
+     * @param headers the headers to be sent to client.
+     */
+    void writeHeaders(Metadata headers);
 
-  protected AbstractServerStream(WritableBufferAllocator bufferAllocator,
-                                 int maxMessageSize) {
-    super(bufferAllocator, maxMessageSize);
+    /**
+     * Sends an outbound frame to the remote end point.
+     *
+     * @param frame a buffer containing the chunk of data to be sent.
+     * @param flush {@code true} if more data may not be arriving soon
+     */
+    void writeFrame(@Nullable WritableBuffer frame, boolean flush);
+
+    /**
+     * Sends trailers to the remote end point. This call implies end of stream.
+     *
+     * @param trailers metadata to be sent to end point
+     * @param headersSent {@code true} if response headers have already been sent.
+     */
+    void writeTrailers(Metadata trailers, boolean headersSent);
+
+    void request(int numMessages);
+
+    void cancel(Status status);
   }
 
-  /**
-   * Sets the listener to receive notifications. Must be called in the context of the transport
-   * thread.
-   */
-  public final void setListener(ServerStreamListener listener) {
-    this.listener = checkNotNull(listener);
+  private final MessageFramer framer;
+  private boolean outboundClosed;
+  private boolean headersSent;
 
-    // Now that the stream has actually been initialized, call the listener's onReady callback if
-    // appropriate.
-    onStreamAllocated();
+  protected AbstractServerStream(WritableBufferAllocator bufferAllocator) {
+    this.framer = new MessageFramer(this, bufferAllocator);
   }
 
   @Override
-  protected ServerStreamListener listener() {
-    return listener;
+  protected abstract TransportState transportState();
+
+  protected abstract Sink abstractServerStreamSink();
+
+  @Override
+  protected MessageFramer framer() {
+    return framer;
   }
 
   @Override
-  protected void receiveMessage(InputStream is) {
-    inboundPhase(Phase.MESSAGE);
-    listener().messageRead(is);
+  public final void request(int numMessages) {
+    abstractServerStreamSink().request(numMessages);
```

Comment:
`AbstractStream` said it was only to be used from the application thread, and that will continue through the whole hierarchy. The same is true for TransportState. You have to understand the base class. Repeating that information would require it 12 additional places (here, Netty server, Netty client, OkHttp client, AbstractClientStream, and Http2ClientStream; times two for stream and transport state). If it were only one or two additional places I could understand, but the amount of repetition necessary in this case seems too much.